doc/man/id3info.1
.TH ID3INFO 1 "May 2000" local "User Command"
.SH NAME
id3info \- Display id3 tag information.
.SH SYNOPSIS
.B id3info
.RB [
.I OPTION
.RB ]
.RB [
.I FILE
.RB ]
.br
.SH DESCRIPTION
.B Id3info
displays both the id3v1 and id3v2 tag information for a file.
Id3info will not differentiate between the two types of tags.
.SH OPTIONS
.TP
.B \-h, \-\-help
Display help and exit
.TP
.B \-v, \-\-version
Display version information and exit
.SH SEE ALSO
id3convert(1), id3tag(1), id3v2(1)
.SH AUTHOR
.B id3lib
was originally designed and implemented by Dirk Mahoney and is
maintained by Scott Thomas Haug <sth2@cs.wustl.edu>. Manual page written for
Debian GNU/Linux by Robert Woodcock <rcw@debian.org>.



ChangeLog
2006-02-17  Jerome Couderc

    * Patch from Spoon to fix UTF-16 writing bug
      http://sourceforge.net/tracker/index.php?func=detail&aid=1016290&group_id=979&atid=300979

2003-03-02 Sunday 17:38   Thijmen Klok <thijmen@id3lib.org>

	* THANKS (1.20): added more people 

2003-03-02 Sunday 17:37   Thijmen Klok <thijmen@id3lib.org>

	* config.h.win32 (1.25), config.h.win32.in (1.13), configure
	(1.79), configure.in (1.68), macconfig.h (1.12): changes to reflect
	new version 

2003-03-02 Sunday 17:19   Thijmen Klok <thijmen@id3lib.org>

	* config.sub (1.3), config.guess (1.3): update from
	ftp://ftp.gnu.org/pub/gnu/config/ timestamp 2003-02-22 

2003-03-02 Sunday 16:17   Thijmen Klok <thijmen@id3lib.org>

	* examples/: demo_main.cpp (1.4), test_io.cpp (1.6),
	test_remove.cpp (1.13): added cerr cout and endl where needed 

2003-03-02 Sunday 15:39   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_find.cpp (1.28): Fixes a bug where a string is tried to
	be made from NULL 

2003-03-02 Sunday 15:30   Thijmen Klok <thijmen@id3lib.org>

	* src/header_tag.cpp (1.25): adds a check on _info for NULL 

2003-03-02 Sunday 15:23   Thijmen Klok <thijmen@id3lib.org>

	* src/: field_string_ascii.cpp (1.29), field_string_unicode.cpp
	(1.33), field_binary.cpp (1.27): added a check for NULL in Set() 

2003-03-02 Sunday 15:14   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_file.cpp (1.43): fixes left open file 

2003-03-02 Sunday 14:56   Thijmen Klok <thijmen@id3lib.org>

	* include/id3/id3lib_strings.h (1.4): fix for undefined wchar_t
	template 

2003-03-02 Sunday 14:35   Thijmen Klok <thijmen@id3lib.org>

	* src/tag.cpp (1.55), include/id3/tag.h (1.63): fix because
	GetFileName need to return a pointer which keeps to be valid 

2003-02-21 Friday 04:47  slackorama

	* include/id3/globals.h (1.54): added a missing const to remove
	warning when compiling with -Wwrite-strings (as is done in KDE and
	many other projects) 

2002-11-24 Sunday 20:39   Thijmen Klok <thijmen@id3lib.org>

	* ChangeLog (1.138): added all entries for 3.8.2 release 

2002-11-24 Sunday 20:34   Thijmen Klok <thijmen@id3lib.org>

	* id3com/id3com.dsp (1.20): added link32 to zlib 

2002-11-24 Sunday 18:33   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_parse.cpp (1.47): fixed a bug which caused v1 tags to be
	ignored after a lyrics v2 tag 

2002-11-24 Sunday 17:31   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_parse.cpp (1.46): fixed a stupid bug of mine, affecting
	wrong handling of id3v1 tags 

2002-11-03 Sunday 01:41   Thijmen Klok <thijmen@id3lib.org>

	* src/field.cpp (1.47): fixes access to invalid framedefs 

2002-11-02 Saturday 23:28   Thijmen Klok <thijmen@id3lib.org>

	* id3com/win32.readme.first.txt (1.4),
	libprj/win32.readme.first.txt (1.4), prj/win32.readme.first.txt
	(1.4): improved comment on delphi 

2002-11-02 Saturday 22:02   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_file.cpp (1.42): fixed a bug in RenderV2ToFile, file was
	not cleared and thus eof mark was still set 

2002-11-02 Saturday 19:10   Thijmen Klok <thijmen@id3lib.org>

	* ChangeLog (1.137): added todays work 

2002-11-02 Saturday 19:07   Thijmen Klok <thijmen@id3lib.org>

	* config.h.win32 (1.24), configure (1.78), configure.in (1.67),
	macconfig.h (1.11): bumped version up to 3.8.1 

2002-11-02 Saturday 19:03   Thijmen Klok <thijmen@id3lib.org>

	* id3lib.spec.in (1.27): added a line on mp3 headers 

2002-11-02 Saturday 18:49   Thijmen Klok <thijmen@id3lib.org>

	* THANKS (1.19): added latest help 

2002-11-02 Saturday 18:48   Thijmen Klok <thijmen@id3lib.org>

	* include/id3/globals.h (1.53), src/mp3_header.h (1.4),
	src/mp3_parse.cpp (1.6): added vbr support 

2002-11-02 Saturday 18:35   Thijmen Klok <thijmen@id3lib.org>

	* include/id3/tag.h (1.62), src/tag.cpp (1.54), src/tag_file.cpp
	(1.41), src/tag_impl.h (1.10), src/tag_parse.cpp (1.45): added
	streaming support 

2002-11-02 Saturday 18:03   Thijmen Klok <thijmen@id3lib.org>

	* libprj/win32.readme.first.txt (1.3),
	id3com/win32.readme.first.txt (1.3), prj/win32.readme.first.txt
	(1.3): added comment about id3lib.pas 

2002-11-02 Saturday 17:56   Thijmen Klok <thijmen@id3lib.org>

	* delphi/Id3lib.pas (1.1): added delphi interface to id3lib 

2002-11-02 Saturday 13:45   Thijmen Klok <thijmen@id3lib.org>

	* id3com/: ID3Tag.cpp (1.14), id3com.dsp (1.19), id3com.idl (1.13): 
	Updated id3com in line with latest id3lib 

2002-10-04 Friday 10:57   Thijmen Klok <thijmen@id3lib.org>

	* ChangeLog (1.136): added latest entries 

2002-10-04 Friday 10:52   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_parse_lyrics3.cpp (1.35): cleaning source 

2002-09-23 Monday 23:32  slackorama

	* src/tag_parse_lyrics3.cpp (1.34): Fixed a "version priority"
	issue in the ID3 lyrics tagging stuff.  Posted patch on id3lib-devel
	first.  

2002-09-21 Saturday 19:23   Thijmen Klok <thijmen@id3lib.org>

	* include/id3.h (1.21), include/id3/tag.h (1.61), src/c_wrapper.cpp
	(1.22, rel-3-7-10), src/helpers.cpp (1.12), src/tag.cpp (1.53),
	src/tag_impl.cpp (1.13), src/tag_impl.h (1.9): added boolean return
	for AttachFrame 

2002-09-21 Saturday 18:04   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_impl.cpp (1.12): initializing _mp3_info to NULL 

2002-09-19 Thursday 12:20   Thijmen Klok <thijmen@id3lib.org>

	* src/misc_support.cpp (1.39), include/id3/misc_support.h (1.29): 
	fixed prototpyes ID3_GetSyncLyrics 

2002-08-13 Tuesday 18:10   Thijmen Klok <thijmen@id3lib.org>

	* include/id3/id3lib_strings.h (1.3): fixed compile for gcc 3.2 

2002-08-11 Sunday 12:58   Thijmen Klok <thijmen@id3lib.org>

	* cvs2cl.pl (1.2): changed command for use of ssh 

2002-08-10 Saturday 22:39   Thijmen Klok <thijmen@id3lib.org>

	* config.h.win32.in (1.12): updated for version info on dll 

2002-08-10 Saturday 22:13   Thijmen Klok <thijmen@id3lib.org>

	* ChangeLog (1.135): updated changelog for latest changes before
	final 

2002-08-10 Saturday 22:02   Thijmen Klok <thijmen@id3lib.org>


examples/id3info.cpp
// SPDX-License-Identifier: LGPL-3.0-only

#include <config.h>
#include <id3/io_helpers.h>
#include <id3/misc_support.h>
#include <id3/readers.h>
#include <id3/tag.h>
#include <id3/utils.h>

#define APP_NAME "id3info"

using namespace dami;
using std::cout;
using std::endl;

/*
 * Replace the gengetopt generated code with basic arg parsing. Even though
 * getopt() is now part of libc, the file was attached here probably for other
 * architectures. In order to support Windows, focus on the main purpose of
 * this utility and parse filenams directly with couple of basic options.
 * One downside is that it's too basic to combine multiple short options in one
 */
void PrintUsage ()
{
    cout << "Display the id3 (both v1 and v2) tag information for one or more files." << endl;
    cout << "Usage: " << APP_NAME << " <FILES>..." << endl;
    cout << "       " << APP_NAME << " [-h | -V]" << endl;
    cout << endl;
    cout << "  -h, --help      Print help and exit" << endl;
    cout << "  -V, --version   Print version and exit" << endl;
    cout << endl;
}

void PrintVersion ()
{
    cout << APP_NAME << " version: " << VERSION << endl;
    cout << "Displays ID3 Tag Information - Written by Scott Thomas Haug" << endl;
    cout << "Uses " << ID3LIB_FULL_NAME << endl;
    cout << endl;
}

void PrintInformation (const ID3_Tag &myTag)
{
    ID3_Tag::ConstIterator *iter = myTag.CreateIterator();
    const ID3_Frame *frame = NULL;

    while (NULL != (frame = iter->GetNext())) {
        const char *desc = frame->GetDescription();

        if (!desc) {
            desc = "";
        }

        cout << "=== " << frame->GetTextID() << " (" << desc << "): ";
        ID3_FrameID eFrameID = frame->GetID();

        switch (eFrameID) {
        case ID3FID_ALBUM:
        case ID3FID_BPM:
        case ID3FID_COMPOSER:
        case ID3FID_CONTENTTYPE:
        case ID3FID_COPYRIGHT:
        case ID3FID_DATE:
        case ID3FID_PLAYLISTDELAY:
        case ID3FID_ENCODEDBY:
        case ID3FID_LYRICIST:
        case ID3FID_FILETYPE:
        case ID3FID_TIME:
        case ID3FID_CONTENTGROUP:
        case ID3FID_TITLE:
        case ID3FID_SUBTITLE:
        case ID3FID_INITIALKEY:
        case ID3FID_LANGUAGE:
        case ID3FID_SONGLEN:
        case ID3FID_MEDIATYPE:
        case ID3FID_ORIGALBUM:
        case ID3FID_ORIGFILENAME:
        case ID3FID_ORIGLYRICIST:
        case ID3FID_ORIGARTIST:
        case ID3FID_ORIGYEAR:
        case ID3FID_FILEOWNER:
        case ID3FID_LEADARTIST:
        case ID3FID_BAND:
        case ID3FID_CONDUCTOR:
        case ID3FID_MIXARTIST:
        case ID3FID_PARTINSET:
        case ID3FID_PUBLISHER:
        case ID3FID_TRACKNUM:
        case ID3FID_RECORDINGDATES:
        case ID3FID_NETRADIOSTATION:
        case ID3FID_NETRADIOOWNER:
        case ID3FID_SIZE:
        case ID3FID_ISRC:
        case ID3FID_ENCODERSETTINGS:
        case ID3FID_YEAR: {
            char *sText = ID3_GetString (frame, ID3FN_TEXT);
            cout << sText << endl;
            delete [] sText;
            break;
        }

        case ID3FID_USERTEXT: {
            char
            *sText = ID3_GetString (frame, ID3FN_TEXT),
             *sDesc = ID3_GetString (frame, ID3FN_DESCRIPTION);
            cout << "(" << sDesc << "): " << sText << endl;
            delete [] sText;
            delete [] sDesc;
            break;
        }

        case ID3FID_COMMENT:
        case ID3FID_UNSYNCEDLYRICS: {
            char
            *sText = ID3_GetString (frame, ID3FN_TEXT),
             *sDesc = ID3_GetString (frame, ID3FN_DESCRIPTION),
              *sLang = ID3_GetString (frame, ID3FN_LANGUAGE);
            cout << "(" << sDesc << ")[" << sLang << "]: "
                 << sText << endl;
            delete [] sText;
            delete [] sDesc;
            delete [] sLang;
            break;
        }

        case ID3FID_WWWAUDIOFILE:
        case ID3FID_WWWARTIST:
        case ID3FID_WWWAUDIOSOURCE:
        case ID3FID_WWWCOMMERCIALINFO:
        case ID3FID_WWWCOPYRIGHT:
        case ID3FID_WWWPUBLISHER:
        case ID3FID_WWWPAYMENT:
        case ID3FID_WWWRADIOPAGE: {
            char *sURL = ID3_GetString (frame, ID3FN_URL);
            cout << sURL << endl;
            delete [] sURL;
            break;
        }

        case ID3FID_WWWUSER: {
            char
            *sURL = ID3_GetString (frame, ID3FN_URL),
             *sDesc = ID3_GetString (frame, ID3FN_DESCRIPTION);
            cout << "(" << sDesc << "): " << sURL << endl;
            delete [] sURL;
            delete [] sDesc;
            break;
        }

        case ID3FID_INVOLVEDPEOPLE: {
            size_t nItems = frame->GetField (ID3FN_TEXT)->GetNumTextItems();

            for (size_t nIndex = 0; nIndex < nItems; nIndex++) {
                char *sPeople = ID3_GetString (frame, ID3FN_TEXT, nIndex);
                cout << sPeople;
                delete [] sPeople;

                if (nIndex + 1 < nItems) {
                    cout << ", ";
                }
            }

            cout << endl;
            break;
        }

        case ID3FID_PICTURE: {
            char
            *sMimeType = ID3_GetString (frame, ID3FN_MIMETYPE),
             *sDesc     = ID3_GetString (frame, ID3FN_DESCRIPTION),
              *sFormat   = ID3_GetString (frame, ID3FN_IMAGEFORMAT);
            size_t
            nPicType   = frame->GetField (ID3FN_PICTURETYPE)->Get(),
            nDataSize  = frame->GetField (ID3FN_DATA)->Size();
            cout << "(" << sDesc << ")[" << sFormat << ", "
                 << nPicType << "]: " << sMimeType << ", " << nDataSize
                 << " bytes" << endl;
            delete [] sMimeType;
            delete [] sDesc;
            delete [] sFormat;
            break;
        }

        case ID3FID_GENERALOBJECT: {
            char
            *sMimeType = ID3_GetString (frame, ID3FN_MIMETYPE),
             *sDesc = ID3_GetString (frame, ID3FN_DESCRIPTION),
              *sFileName = ID3_GetString (frame, ID3FN_FILENAME);
            size_t
            nDataSize = frame->GetField (ID3FN_DATA)->Size();
            cout << "(" << sDesc << ")["
                 << sFileName << "]: " << sMimeType << ", " << nDataSize
                 << " bytes" << endl;
            delete [] sMimeType;
            delete [] sDesc;
            delete [] sFileName;
            break;
        }

        case ID3FID_UNIQUEFILEID: {
            char *sOwner = ID3_GetString (frame, ID3FN_OWNER);
            size_t nDataSize = frame->GetField (ID3FN_DATA)->Size();
            cout << sOwner << ", " << nDataSize
                 << " bytes" << endl;
            delete [] sOwner;
            break;
        }

        case ID3FID_PLAYCOUNTER: {
            size_t nCounter = frame->GetField (ID3FN_COUNTER)->Get();
            cout << nCounter << endl;
            break;
        }

        case ID3FID_POPULARIMETER: {
            char *sEmail = ID3_GetString (frame, ID3FN_EMAIL);
            size_t
            nCounter = frame->GetField (ID3FN_COUNTER)->Get(),
            nRating = frame->GetField (ID3FN_RATING)->Get();
            cout << sEmail << ", counter="
                 << nCounter << " rating=" << nRating << endl;
            delete [] sEmail;
            break;
        }

        case ID3FID_CRYPTOREG:
        case ID3FID_GROUPINGREG: {
            char *sOwner = ID3_GetString (frame, ID3FN_OWNER);
            size_t
            nSymbol = frame->GetField (ID3FN_ID)->Get(),
            nDataSize = frame->GetField (ID3FN_DATA)->Size();
            cout << "(" << nSymbol << "): " << sOwner
                 << ", " << nDataSize << " bytes" << endl;
            break;
        }

        case ID3FID_SYNCEDLYRICS: {
            char
            *sDesc = ID3_GetString (frame, ID3FN_DESCRIPTION),
             *sLang = ID3_GetString (frame, ID3FN_LANGUAGE);
            size_t
            nTimestamp = frame->GetField (ID3FN_TIMESTAMPFORMAT)->Get(),
            nRating = frame->GetField (ID3FN_CONTENTTYPE)->Get();
            const char *format = (2 == nTimestamp) ? "ms" : "frames";
            cout << "(" << sDesc << ")[" << sLang << "]: ";

            switch (nRating) {
            case ID3CT_OTHER:
                cout << "Other";
                break;

            case ID3CT_LYRICS:
                cout << "Lyrics";
                break;

            case ID3CT_TEXTTRANSCRIPTION:
                cout << "Text transcription";
                break;

            case ID3CT_MOVEMENT:
                cout << "Movement/part name";
                break;

            case ID3CT_EVENTS:
                cout << "Events";
                break;

            case ID3CT_CHORD:
                cout << "Chord";
                break;

            case ID3CT_TRIVIA:
                cout << "Trivia/'pop up' information";
                break;
            }

            cout << endl;
            ID3_Field *fld = frame->GetField (ID3FN_DATA);

            if (fld) {
                ID3_MemoryReader mr (fld->GetRawBinary(), fld->BinSize());

                while (!mr.atEnd()) {
                    cout << io::readString (mr).c_str();
                    cout << " [" << io::readBENumber (mr, sizeof (uint32)) << " "
                         << format << "] ";
                }
            }

            cout << endl;
            delete [] sDesc;
            delete [] sLang;
            break;
        }

        case ID3FID_AUDIOCRYPTO:
        case ID3FID_EQUALIZATION:
        case ID3FID_EVENTTIMING:
        case ID3FID_CDID:
        case ID3FID_MPEGLOOKUP:
        case ID3FID_OWNERSHIP:
        case ID3FID_PRIVATE:
        case ID3FID_POSITIONSYNC:
        case ID3FID_BUFFERSIZE:
        case ID3FID_VOLUMEADJ:
        case ID3FID_REVERB:
        case ID3FID_SYNCEDTEMPO:
        case ID3FID_METACRYPTO: {
            cout << " (unimplemented)" << endl;
            break;
        }

        default: {
            cout << " frame" << endl;
            break;
        }
        }
    }

    delete iter;
}

int main (int argc, char *const argv[])
{
    ID3D_INIT_DOUT();
    ID3D_INIT_WARNING();
    ID3D_INIT_NOTICE();

    if (argc < 2) {
        PrintUsage();
        exit (1);
    }

    std::string arg1 = argv[1];

    if (arg1 == "-h" || arg1 == "--help") {
        PrintUsage();
        exit (0);
    } else if (arg1 == "-V" || arg1 == "--version") {
        PrintVersion();
        exit (0);
    } else if (arg1.at (0) == '-') {
        cout << "Error, Invalid option: " << arg1 << endl << endl;
        PrintUsage();
        exit (1);
    }

    std::string filename;

    for (size_t i = 1; i < argc; i++) {
        filename = argv[i];

        ID3_Tag myTag;
        myTag.Link (filename.c_str(), ID3TT_ALL);
        const Mp3_Headerinfo *mp3info;
        mp3info = myTag.GetMp3HeaderInfo();

        cout << endl << "*** Tag information for " << filename << endl;
        PrintInformation (myTag);

        if (mp3info) {
            cout << "*** mp3 info\n";

            switch (mp3info->version) {
            case MPEGVERSION_2_5:
                cout << "MPEG2.5/";
                break;

            case MPEGVERSION_2:
                cout << "MPEG2/";
                break;

            case MPEGVERSION_1:
                cout << "MPEG1/";
                break;

            default:
                break;
            }

            switch (mp3info->layer) {
            case MPEGLAYER_III:
                cout << "layer III\n";
                break;

            case MPEGLAYER_II:
                cout << "layer II\n";
                break;

            case MPEGLAYER_I:
                cout << "layer I\n";
                break;

            default:
                break;
            }

            cout << "Bitrate: " << mp3info->bitrate / 1000 << "KBps\n";
            cout << "Frequency: " << mp3info->frequency / 1000 << "KHz\n";
        }


    }

    return 0;
}
